kernel/kernel-api/src/main/java/io/delta/kernel/internal/tablefeatures/TableFeatures.java

```diff
@@ -0,0 +1,693 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.kernel.internal.tablefeatures;
+
+import static io.delta.kernel.internal.DeltaErrors.*;
+import static io.delta.kernel.internal.TableConfig.APPEND_ONLY;
+import static io.delta.kernel.internal.TableConfig.COLUMN_MAPPING_MODE;
+import static io.delta.kernel.internal.TableConfig.IN_COMMIT_TIMESTAMPS_ENABLED;
+import static io.delta.kernel.internal.TableConfig.ROW_TRACKING_ENABLED;
+import static io.delta.kernel.internal.util.ColumnMapping.ColumnMappingMode.NONE;
+
+import io.delta.kernel.internal.DeltaErrors;
+import io.delta.kernel.internal.TableConfig;
+import io.delta.kernel.internal.actions.Metadata;
+import io.delta.kernel.internal.actions.Protocol;
+import io.delta.kernel.internal.util.CaseInsensitiveMap;
+import io.delta.kernel.internal.util.ColumnMapping;
+import io.delta.kernel.internal.util.Tuple2;
+import io.delta.kernel.types.StructType;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/** Contains utility methods related to the Delta table feature support in protocol. */
+public class TableFeatures {
+
+  /** Min reader version that supports reader features. */
+  public static final int TABLE_FEATURES_MIN_READER_VERSION = 3;
+
+  /** Min reader version that supports writer features. */
+  public static final int TABLE_FEATURES_MIN_WRITER_VERSION = 7;
+
+  //////////////////////////////////////////////////////////////////////
+  /// Define the {@link TableFeature}s that are supported by Kernel  ///
+  //////////////////////////////////////////////////////////////////////
+  /** An interface to indicate a feature is legacy, i.e., released before Table Features. */
+  public interface LegacyFeatureType extends FeatureAutoEnabledByMetadata {
+    @Override
+    default boolean automaticallyUpdateProtocolOfExistingTables() {
+      return true; // legacy features can always be applied to existing tables
+    }
+  }
+
+  /** An interface to indicate a feature applies to readers and writers. */
+  public interface ReaderWriterFeatureType {}
+
+  /** A base class for all table legacy writer-only features. */
+  public abstract static class LegacyWriterFeature extends TableFeature
+      implements LegacyFeatureType {
+    public LegacyWriterFeature(String featureName, int minWriterVersion) {
+      super(featureName, /* minReaderVersion = */ 0, minWriterVersion);
+    }
+  }
+
+  /** A base class for all table legacy reader-writer features. */
+  public abstract static class LegacyReaderWriterFeature extends TableFeature
```

Comment:
should this implement `ReaderWriterFeatureType`?

Reply:
I think it must? Otherwise `isReaderWriterFeature` is wrong?

Reply:
yep, added it.